Document/0x06f-Testing-Local-Authentication.md

```diff
@@ -9,26 +9,196 @@ As a result only requirement "4.6	Biometric authentication, if any, is not event
+	guard success else {
+		// User did not authenticate successfully, look at evaluationError and take appropriate action
+	}
+
+	// User authenticated successfully, take appropriate action
+}
+```
+
+###### Objective-C
+```
+LAContext *myContext = [[LAContext alloc] init];
+NSError *authError = nil;
+
+if ([myContext canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
+    [myContext evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
+                  localizedReason:@"Please, pass authorisation to enter this area"
+                            reply:^(BOOL success, NSError *error) {
+            if (success) {
+                // User authenticated successfully, take appropriate action
+            } else {
+                // User did not authenticate successfully, look at error and take appropriate action
+            }
+        }];
+} else {
+    // Could not evaluate policy; look at authError and present an appropriate message to user
+}
+```
+
+##### Security.framework
+Keychain allows to save items with special SecAccessControl attribute, which will allow to get item from Keychain only after user will pass Touch ID authentication (or passcode, if such fallback is allowed by attribute parameters).
+
+
+
+In following example we will save "S00p3r_haCk3r_strong_password" string to Keychain which can be accessed only on current device while passcode is set (`kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` parameter) after Touch ID authentication for currently enrolled fingers only (`.touchIDCurrentSet parameter`):
+###### Swift
+```
+// 1. create AccessControl object that will represent authentication settings
+
+var error: Unmanaged<CFError>?
+
+guard let accessControl = SecAccessControlCreateWithFlags(kCFAllocatorDefault,
+	kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
+	.touchIDCurrentSet,
+	&error) else {
+    // failed to create AccessControl object
+}
+
+// 2. define Keychain services query. Pay attention that kSecAttrAccessControl is mutually exclusive with kSecAttrAccessible attribute
+
+var query: Dictionary<String, Any> = [:]
+
+query[kSecClass as String] = kSecClassGenericPassword
+query[kSecAttrLabel as String] = "com.me.myapp.password" as CFString
+query[kSecAttrAccount as String] = "OWASP Account" as CFString
+query[kSecValueData as String] = "S00p3r_haCk3r_strong_password".data(using: .utf8)! as CFData
+query[kSecAttrAccessControl as String] = accessControl
+
+// 3. save item
+
+let status = SecItemAdd(query as CFDictionary, nil)
+
+if status == noErr {
+	// successfully saved
+} else {
+	// error while saving
+}
+```
+###### Objective-C
+```
+// 1. create AccessControl object that will represent authentication settings
+CFErrorRef *err = nil;
+
+SecAccessControlRef sacRef = SecAccessControlCreateWithFlags(kCFAllocatorDefault,
+	kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
+	kSecAccessControlUserPresence,
```

Comment:
this is an entire different access item:) and is actually the weakest. Using this, will not fully enable the secure-enclave in the same way using the 2 touchID related tokens do. See key management for more details. 

In fact, I would rather expect in static analysis that the pentester can verify that LAContext is not used, that this token is not used and that using the keychain-based entries are the proper remediation :).